pkg/miscellaneous/miscellaneous.go

```diff
@@ -520,3 +522,31 @@ func ParsePeerURL(initialAdvertisePeerURLs, podName string) (string, error) {
 	domaiName := fmt.Sprintf("%s.%s.%s", tokens[1], tokens[2], "svc")
 	return fmt.Sprintf("%s://%s.%s:%s", tokens[0], podName, domaiName, tokens[3]), nil
 }
+
+// IsPeerURLTLSEnabled checks whether the peer address is TLS enabled or not.
+func IsPeerURLTLSEnabled() (bool, error) {
```

Comment:
This is not sufficient the last time i checked. Since CM is mounted etcd-backup-restore will get an updated configuration which only captures the desired state as set in the etcd resource. However this does not correctly reflect the state of the actual/current peer URL TLS protocol which can only be fetched by checking the member peer URL.

Reply:
I'm also not doing anything different from the current code. Let me explain:
Initially I also thought that we are getting peerURL using some api call to Etcd but later I found that's not the case.
If you look at the current code [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L208-L218), it is calling [UpdateMemberPeerURL](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/member/member_control.go#L246) to get peerURL, which calls [doUpdateMemberPeerAddress](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/member/member_control.go#L188) which is getting the peerURL from configMap check [here](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/member/member_control.go#L174) , so I felt there is no need for any api call required just to check the TLS is enabled or not.
So, I separated the functionality of updating the PeerURL and check the TLS is enabled or not.

Reply:
There is a slight difference. In [doUpdateMemberPeerAddress](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/member/member_control.go#L188-L206) the member peer URL read from the config is returned only when the same URL is successfully updated by calling `cli.MemberUpdate(memberUpdateCtx, id, []string{memberPeerURL})` else an error is returned.

Reply:
> cli.MemberUpdate(memberUpdateCtx, id, []string{memberPeerURL})

this api call will never be successful if restoration of etcd will be going on, then [doUpdateMemberPeerAddress](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/member/member_control.go#L188-L206) will return empty string which leads to this issue.
Here, I'm not using `PeerURL`  for anything apart from checking whether it is `http` or `https` .
Can you give me the scenario of this edge where this can cause issue as may be I don't have much of a insight on this whole TLS not enabled -> TLS enabled case scenario ?

Reply:
So let us assume that the configuration has been updated and as a result the ConfigMap change reaches backup-restore. The change is that peer URL has been TLS enabled. In etcd-controller (etcd-druid), the configmap got updated but due to an error the event was re-queued. In the mean time backup-restore got the new config. In your code above this value was read and on the lease it was set as true against the annotation.

Now when etcd-controller finally services the re-queued event then it will find that this has been enabled on the member and will do nothing when it should have actually deleted the STS.

Reply:
If that's the case then this edge is also applicable to the current code. IMO, current code is also not handling this edge case. If I'm wrong then please explain how's the current code handling this edge case.
Nevertheless, I would suggest to open new issue with whatever you have describe above as handling of this edge case is not in scope of this PR.

Reply:
Various edge cases I have described here: https://github.com/gardener/etcd-backup-restore/pull/564#discussion_r1070760238

Reply:
> In your code above this value was read and on the lease it was set as true against the annotation.

actually no, lease annotation will still be false as we are checking TLSEnable case only on startup (in current code as well as in this PR) and passing the boolean variable to wherever it requires.